"Cannot find module '@radix-ui/react-dialog'" on a new project

### Minimal reproducible example

https://github.com/Gelio/expo-router-radix-ui-react-modal-bug

### Steps to reproduce

The https://github.com/Gelio/expo-router-radix-ui-react-modal-bug repository is a result of running `npx create-expo-app@latest` with the `default` template, and then immediately running `npm run web`. `@radix-ui/react-dialog` cannot be found in `node_modules`, for some reason.

1. `npm install`
2. `npm run web`

<details>

<summary>Logs</summary>

```sh
at 16:43:15 ❯ npm run web

> kwiapp@1.0.0 web
> expo start --web

Starting project at /Users/grzegorz.rozdzialik/projects/personal/kwiapp
React Compiler enabled
Starting Metro Bundler
▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
█ ▄▄▄▄▄ █▀▄█▀ ██  █ ▄▄▄▄▄ █
█ █   █ █▄   ▄█▄▄██ █   █ █
█ █▄▄▄█ █ ▀█▀█▀▀ ██ █▄▄▄█ █
█▄▄▄▄▄▄▄█ ▀▄█ █ ▀ █▄▄▄▄▄▄▄█
█ ▄█ ▀█▄▄██ ▀█▄▀██▀  ▄▀▄▄▀█
█ ▀█▀▀█▄█▀▄▄██ ▄▀ ▄▀ ▀▀█▄▄█
██ █▀▄▀▄█▀▀▀▀   █▀█ ▄█ ██▀█
█▄▀▀ █▀▄█ ▀ █  ██ ▄▄ ▀▀██▄█
█▄▄▄█▄▄▄█▀█ ▄▄█▄  ▄▄▄ █ ▄ █
█ ▄▄▄▄▄ █▄▀ ▀██▄  █▄█  ▀▄▄█
█ █   █ █▀▄▀▄█ ▀▀▄ ▄▄ █▀▄██
█ █▄▄▄█ █▀▀▀█▄  █  █▄  ▄█▄█
█▄▄▄▄▄▄▄█▄▄▄▄█▄▄█▄███▄▄█▄▄█

› Metro waiting on exp://192.168.50.199:8081
› Scan the QR code above with Expo Go (Android) or the Camera app (iOS)

› Web is waiting on http://localhost:8081

› Using Expo Go
› Press s │ switch to development build

› Press a │ open Android
› Press i │ open iOS simulator
› Press w │ open web

› Press j │ open debugger
› Press r │ reload app
› Press m │ toggle menu
› shift+m │ more tools
› Press o │ open project code in your editor

› Press ? │ show all commands

Logs for your project will appear below. Press Ctrl+C to exit.
λ Bundled 665ms node_modules/expo-router/node/render.js (1072 modules)
Web node_modules/expo-router/entry.js ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░ 99.9% (1204/1204)
TypeError: The argument 'path' must be a string, Uint8Array, or URL without null bytes. Received '/Users/grzegorz.rozdzialik/projects/personal/\x00node:@radix-ui/react-dialog'
    at Object.readFileSync (node:fs:439:14)
    at getCodeFrame (/Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/metro/src/Server.js:997:18)
    at Server._symbolicate (/Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/metro/src/Server.js:1079:22)
    at Server._processRequest (/Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/metro/src/Server.js:460:7) {
  code: 'ERR_INVALID_ARG_VALUE'

Metro error: Cannot find module '@radix-ui/react-dialog'
Require stack:
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo-router/node/render.js.bundle?platform=web&dev=true&hot=false&transform.engine=hermes&transform.routerRoot=app&resolver.environment=node&transform.environment=node&unstable_transformProfile=hermes-stable
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/server/getStaticRenderFunctions.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/server/metro/metroErrorInterface.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/server/metro/createServerComponentsMiddleware.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/server/metro/MetroBundlerDevServer.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/server/DevServerManager.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/startAsync.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/src/start/index.js
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/node_modules/@expo/cli/build/bin/cli
- /Users/grzegorz.rozdzialik/projects/personal/kwiapp/node_modules/expo/bin/cli

Call Stack
  Function._resolveFilename (node:internal/modules/cjs/loader)
  defaultResolveImpl (node:internal/modules/cjs/loader)
  resolveForCJSWithHooks (node:internal/modules/cjs/loader)
  Function._load (node:internal/modules/cjs/loader)
  TracingChannel.traceSync (node:diagnostics_channel)
  wrapModuleLoad (node:internal/modules/cjs/loader)
  Module.require (node:internal/modules/cjs/loader)
  $$require_external (node:internal/modules/helpers)
  factory (../node:@radix-ui/react-dialog:1:16)
Web Bundled 1294ms node_modules/expo-router/entry.js (1204 modules)
Web Bundled 80ms node_modules/expo-router/_error.js (1 module)
 LOG  [web] Logs will appear in the browser console
```

</details>

The app works fine after I run `npm install @radix-ui/react-dialog` and rerun `npm run web`.

### Environment

```text
expo-env-info 2.0.7 environment info:
    System:
      OS: macOS 15.7.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.19.0 - ~/.local/share/mise/installs/node/22.19.0/bin/node
      Yarn: 1.22.19 - ~/.local/share/mise/installs/node/22.19.0/bin/yarn
      npm: 11.6.1 - ~/.local/share/mise/installs/node/22.19.0/bin/npm
    IDEs:
      Xcode: /undefined - /usr/bin/xcodebuild
    npmPackages:
      expo: ~54.0.12 => 54.0.12
      expo-router: ~6.0.10 => 6.0.10
      react: 19.1.0 => 19.1.0
      react-dom: 19.1.0 => 19.1.0
      react-native: 0.81.4 => 0.81.4
      react-native-web: ~0.21.0 => 0.21.1
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

```text
17/17 checks passed. No issues detected!
```

--- a/packages/@expo/cli/CHANGELOG.md
+++ b/packages/@expo/cli/CHANGELOG.md
@@ -14,6 +14,7 @@
 
 - fix launching apps on Android emulators where `hw.keyboard = no` ([#34677](https://github.com/expo/expo/pull/34677) by [@DimitarNestorov](https://github.com/DimitarNestorov))
 - only skip dependency validation for `EXPO_NO_DEPENDENCY_VALIDATION=1` ([#40043](https://github.com/expo/expo/pull/40043) by [@kitten](https://github.com/kitten))
+- Prevent externals logic from being applied to unreachable Node modules ([#40247](https://github.com/expo/expo/pull/40247) by [@kitten](https://github.com/kitten))
 
 ### 💡 Others
 

--- a/packages/@expo/cli/src/start/server/metro/__tests__/withMetroMultiPlatform.test.ts
+++ b/packages/@expo/cli/src/start/server/metro/__tests__/withMetroMultiPlatform.test.ts
@@ -800,6 +800,12 @@ describe(withExtendedResolver, () => {
             'inline-style-prefixer/index.js',
           ].forEach((name) => {
             it(`externs ${name} to virtual node shim`, () => {
+              jest.mocked(getResolveFunc()).mockImplementation((context, moduleName, _platform) => {
+                return context.originModulePath === '/root/package.json'
+                  ? { type: 'sourceFile', filePath: `mock:${moduleName}` }
+                  : { type: 'empty' };
+              });
+
               const result = config.resolver.resolveRequest!(
                 // Context
                 getNodeResolverContext(),
@@ -815,11 +821,17 @@ describe(withExtendedResolver, () => {
                 `\0node:${name}`
               );
 
-              expect(getResolveFunc()).toHaveBeenCalledTimes(0);
+              expect(getResolveFunc()).toHaveBeenCalledTimes(1);
             });
           });
 
           it(`externs @babel/runtime/xxx subpaths `, () => {
+            jest.mocked(getResolveFunc()).mockImplementation((context, moduleName, _platform) => {
+              return context.originModulePath === '/root/package.json'
+                ? { type: 'sourceFile', filePath: `mock:${moduleName}` }
+                : { type: 'empty' };
+            });
+
             const result = config.resolver.resolveRequest!(
               getNodeResolverContext(),
               '@babel/runtime/xxx/foo.js',
@@ -832,7 +844,7 @@ describe(withExtendedResolver, () => {
               '\0node:@babel/runtime/xxx/foo.js'
             );
 
-            expect(getResolveFunc()).toHaveBeenCalledTimes(0);
+            expect(getResolveFunc()).toHaveBeenCalledTimes(1);
           });
         });
       });

--- a/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
+++ b/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
@@ -248,6 +248,9 @@ export function withExtendedResolver(
     return _universalAliases;
   }
 
+  // used to resolve externals in `requestCustomExternals` from the project root
+  const projectRootOriginPath = path.join(config.projectRoot, 'package.json');
+
   const preferredMainFields: { [key: string]: string[] } = {
     // Defaults from Expo Webpack. Most packages using `react-native` don't support web
     // in the `react-native` field, so we should prefer the `browser` field.
@@ -519,7 +522,7 @@ export function withExtendedResolver(
 
     // Custom externals support
     function requestCustomExternals(
-      context: ResolutionContext,
+      immutableContext: ResolutionContext,
       moduleName: string,
       platform: string | null
     ) {
@@ -528,12 +531,22 @@ export function withExtendedResolver(
         return null;
       }
       // Skip applying JS externals for CSS files.
-      if (/\.(s?css|sass)$/.test(context.originModulePath)) {
+      if (/\.(s?css|sass)$/.test(immutableContext.originModulePath)) {
         return null;
       }
 
-      const environment = context.customResolverOptions?.environment;
+      // TODO(@kitten): Temporary workaround. Our externals logic here isn't generic and only works
+      // for development and not exports. We never intend to use it in exported production bundles,
+      // however, this is still a dangerous implementation. To protect us from externalizing modules
+      // that aren't available to the app, we force any resolution to happen via the project root
+      const context: ResolutionContext = {
+        ...immutableContext,
+        nodeModulesPaths: [],
+        originModulePath: projectRootOriginPath,
+        disableHierarchicalLookup: false,
+      };
 
+      const environment = context.customResolverOptions?.environment;
       const strictResolve = getStrictResolver(context, platform);
 
       for (const external of externals) {
@@ -546,12 +559,15 @@ export function withExtendedResolver(
           } else if (external.replace === 'weak') {
             // TODO: Make this use require.resolveWeak again. Previously this was just resolving to the same path.
             const realModule = strictResolve(moduleName);
-            const realPath = realModule.type === 'sourceFile' ? realModule.filePath : moduleName;
+            const realPath = realModule.type === 'sourceFile' ? realModule.filePath : null;
+            if (!realPath) {
+              return null;
+            }
+
             const opaqueId = idFactory(realPath, {
               platform: platform!,
               environment,
             });
-
             const contents =
               typeof opaqueId === 'number'
                 ? `module.exports=/*${moduleName}*/__r(${opaqueId})`
@@ -569,6 +585,10 @@ export function withExtendedResolver(
               filePath: virtualModuleId,
             };
           } else if (external.replace === 'node') {
+            const externModule = strictResolve(moduleName);
+            if (externModule.type !== 'sourceFile') {
+              return null;
+            }
             const contents = `module.exports=$$require_external('${moduleName}')`;
             const virtualModuleId = `\0node:${moduleName}`;
             debug('Virtualizing Node.js (custom):', moduleName, '->', virtualModuleId);
@@ -581,9 +601,7 @@ export function withExtendedResolver(
               filePath: virtualModuleId,
             };
           } else {
-            throw new CommandError(
-              `Invalid external alias type: "${external.replace}" for module "${moduleName}" (platform: ${platform}, originModulePath: ${context.originModulePath})`
-            );
+            external.replace satisfies never;
           }
         }
       }

--- a/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
+++ b/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
@@ -522,7 +522,7 @@ export function withExtendedResolver(
 
     // Custom externals support
     function requestCustomExternals(
-      immutableContext: ResolutionContext,
+      context: ResolutionContext,
       moduleName: string,
       platform: string | null
     ) {
@@ -531,23 +531,9 @@ export function withExtendedResolver(
         return null;
       }
       // Skip applying JS externals for CSS files.
-      if (/\.(s?css|sass)$/.test(immutableContext.originModulePath)) {
+      if (/\.(s?css|sass)$/.test(context.originModulePath)) {
         return null;
       }
-
-      // TODO(@kitten): Temporary workaround. Our externals logic here isn't generic and only works
-      // for development and not exports. We never intend to use it in exported production bundles,
-      // however, this is still a dangerous implementation. To protect us from externalizing modules
-      // that aren't available to the app, we force any resolution to happen via the project root
-      const context: ResolutionContext = {
-        ...immutableContext,
-        nodeModulesPaths: [],
-        originModulePath: projectRootOriginPath,
-        disableHierarchicalLookup: false,
-      };
-
-      const environment = context.customResolverOptions?.environment;
-      const strictResolve = getStrictResolver(context, platform);
 
       for (const external of externals) {
         if (external.match(context, moduleName, platform)) {
@@ -558,15 +544,11 @@ export function withExtendedResolver(
             };
           } else if (external.replace === 'weak') {
             // TODO: Make this use require.resolveWeak again. Previously this was just resolving to the same path.
-            const realModule = strictResolve(moduleName);
-            const realPath = realModule.type === 'sourceFile' ? realModule.filePath : null;
-            if (!realPath) {
-              return null;
-            }
-
+            const realModule = getStrictResolver(context, platform)(moduleName);
+            const realPath = realModule.type === 'sourceFile' ? realModule.filePath : moduleName;
             const opaqueId = idFactory(realPath, {
               platform: platform!,
-              environment,
+              environment: context.customResolverOptions?.environment,
             });
             const contents =
               typeof opaqueId === 'number'
@@ -585,7 +567,17 @@ export function withExtendedResolver(
               filePath: virtualModuleId,
             };
           } else if (external.replace === 'node') {
-            const externModule = strictResolve(moduleName);
+            // TODO(@kitten): Temporary workaround. Our externals logic here isn't generic and only works
+            // for development and not exports. We never intend to use it in exported production bundles,
+            // however, this is still a dangerous implementation. To protect us from externalizing modules
+            // that aren't available to the app, we force any resolution to happen via the project root
+            const projectRootContext: ResolutionContext = {
+              ...context,
+              nodeModulesPaths: [],
+              originModulePath: projectRootOriginPath,
+              disableHierarchicalLookup: false,
+            };
+            const externModule = getStrictResolver(projectRootContext, platform)(moduleName);
             if (externModule.type !== 'sourceFile') {
               return null;
             }

